addons/block_code/simple_nodes/simple_character/simple_character.gd

```diff
@@ -83,24 +113,46 @@ func _ready():
 
 func simple_setup():
 	add_to_group("affected_by_gravity", true)
+	_setup_actions()
 	_texture_updated()
 
 
-func get_custom_class():
-	return "SimpleCharacter"
+func _setup_actions():
+	if Engine.is_editor_hint() or InputMap.has_action("player_1_left"):
+		return
+
+	for player in PLAYER_KEYS:
+		for action in PLAYER_KEYS[player]:
+			var action_name = player + "_" + action
+			InputMap.add_action(action_name)
+
+			var device = int(player.split("_")[-1]) - 1
 
+			#keyboard event
+			var e = InputEventKey.new()
+			e.physical_keycode = PLAYER_KEYS[player][action]
+			InputMap.action_add_event(action_name, e)
 
-func _player_input_to_direction(player: String):
-	var direction = Vector2()
-	direction.x += float(Input.is_physical_key_pressed(PLAYER_KEYS[player]["right"]))
-	direction.x -= float(Input.is_physical_key_pressed(PLAYER_KEYS[player]["left"]))
-	direction.y += float(Input.is_physical_key_pressed(PLAYER_KEYS[player]["down"]))
-	direction.y -= float(Input.is_physical_key_pressed(PLAYER_KEYS[player]["up"]))
-	return direction
+			#controller d-pad event
+			var ej = InputEventJoypadButton.new()
+			ej.device = device
+			ej.button_index = PLAYER_JOYSTICK_BUTTONS[action]
+			InputMap.action_add_event(action_name, ej)
+
+			#controller left stick event
+			var ejm = InputEventJoypadMotion.new()
+			ejm.device = device
+			ejm.axis = PLAYER_JOYSTICK_MOTION[action]["axis"]
+			ejm.axis_value = PLAYER_JOYSTICK_MOTION[action]["axis_value"]
+			InputMap.action_add_event(action_name, ejm)
+
+
+func get_custom_class():
+	return "SimpleCharacter"
 
 
 func move_with_player_buttons(player: String, kind: String, delta: float):
-	var direction = _player_input_to_direction(player)
+	var direction = Input.get_vector(player + "_left", player + "_right", player + "_up", player + "_down")
```

Comment:
In the `"platformer"` block below, `PLAYER_KEYS` is still hardcoded – you need to adjust this to check the up action instead. Please forgive me missing this in previous rounds of review.

Reply:
Annoyingly there isn't a platformer example in this repo. There is one in https://github.com/endlessm/godot-block-coding-demos, but you can also test it in the pong example by adjusting the block code as follows:

![image](https://github.com/user-attachments/assets/04e4bce8-ae63-4507-a21c-cafbb39931b6)

(It is the "set speed to Vector2(0, 1000)" which prevents the paddles from moving left and righ.t)

Reply:
Thanks for the tip @wjt.

I just tested the platformer code and it should work now. I replaced the hardcoded `PLAYER_KEYS` with `direction.y`

Relevant commit: fd7db2b